index.rst

```diff
@@ -1,59 +1,367 @@
+
+Finally, we propose that all raw ingestion into the shared repository be done with symbolic links or in-place (outside-root) ingests, pointing to read-only filesystems - the existing ``/lsstdata`` for Rubin Observatory data, and a new ``/external-raw`` filesystem for raw data from other instruments.
+
+.. note ::
+
+   **TODO**: This document should probably say more about staging locations for (at least) raws that haven't been ingested yet, for both Rubin Observatory instruments and precursor instruments (which I imagine might be quite different in this respect).  I'd like to at least be confident that the filesystem locations and permissions I'm proposing are consistent with how data arrives and gets ingested.  I don't know enough to write that up myself.
+
+.. _access-controls:
+
+Access Controls
+===============
+
+The current Gen3 registry architecture does not allow any fine-grained access control in the repository database; we instead rely on "friendly users" being careful and respectful of this shared space.
+
+At the same time, we will use filesystem access controls to protect shared and per-user files, and we plan to implement some checks in the Butler client itself to make it at least extremely difficult to *accidentally* cause problems.
+
+.. warning::
+
+   NEVER use ``psql`` or other direct-SQL clients (e.g. the Python DBAPI or SQLAlchemy) to perform write operations in the repository database.
+   These can corrupt the data repository, and we have essentially no way to guard against them.
+
+   It should not be necessary in the long term to ever use direct SQL access even for read access; the SQL schema is *not* considered a public interface - but we recognize that this may be necessary for debugging for a while.
+   This can be ensured by running::
+
+      SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;
+
+   at the start of the session.
+
+   If you have to do this (and not at the prompting of a middleware team member trying to help diagnose a problem), please also create a ticket explaining what you wanted to do that couldn't be done with butler tools, so we can address that feature gap.
+
+This proposal specifies filesystem access controls in terms of a number of high-level "roles" that certain operators or developers may temporarily opt in to via ``su`` or special setuid tools.
+Usually these roles will be used only to create subdirectories that are owned directly by the true user.
+How to map these to users, groups, and filesystem, directory, or file-level permissions in detail is something I'd prefer to leave to the system administrators.
+All directories in ``/datasets`` will be world-readable.
+
+Within each ``repo_<YYYYMMDD>`` repository directory:
+
+ - Regular users will always have write access to their own ``u/<user>`` directory.
```

Comment:
Do users get a `u/<user>` automatically created? Do we automatically create one when a new user gets an account?

Reply:
I think I wrote later in the document that I'd really like the answer to be "yes".

Reply:
Would automatic creation mean that all public repositories have a huge number of `u/` collections (`/project/` has 102 subdirectories, though not all correspond to physical persons)? That might make it hard to get one's bearings when querying collections.

Reply:
I think it's easier to ensure that a `u/` directory exists and is world-writable (mode 1777 like `/tmp`), allowing users to create `u/user/` subdirectories as the Butler needs them, rather than pre-creating them for every user.  (That wouldn't be compatible with the proposal to symlink these into `/project/user`, however.)